runtime/oti/VMHelpers.hpp

```diff
@@ -2048,13 +2049,60 @@ class VM_VMHelpers
 #endif /* JAVA_SPEC_VERSION > 11 */
 	}
 
+#if JAVA_SPEC_VERSION >= 19
+/*
+ *	If low tagging failed (or not even attempted) due to either
+ *
+ *   a carrier thread winning to mount, we don't need to do anything, since it will be compensated by pre/post mount actions
+ *   another GC thread winning to scan, again don't do anything, and let the winning thread do the work, instead
+ */
+	static VMINLINE bool
+	tryEnterConcurrentContinuationMonitor(J9VMContinuation *continuation)
+	{
+		volatile uintptr_t *localAddr = (volatile uintptr_t *) &continuation->carrierThreadAndConcurrentlyScanned;
+		uintptr_t oldValue = *localAddr;
+		uintptr_t retValue = 0;
+
+		if (0 == oldValue) {
+			/* atomically low tag ('set', not 'or') continuation->carrierThreadAndConcurrentlyScanned */
+			if ((retValue = VM_AtomicSupport::lockCompareExchange(localAddr, oldValue, J9_GC_CONTINUATION_CONCURRENTSCANNING)) == oldValue) {
+				return true;
+			}
+		}
+		return false;
+	}
+
+	static VMINLINE void
+	exitConcurrentContinuationMonitor(J9VMContinuation *continuation)
+	{
+		volatile uintptr_t *localAddr = (volatile uintptr_t *) &continuation->carrierThreadAndConcurrentlyScanned;
+
+		uintptr_t oldValue = VM_AtomicSupport::bitAnd(localAddr, ~(uintptr_t)J9_GC_CONTINUATION_CONCURRENTSCANNING);
+		if (J9_ARE_ANY_BITS_SET(oldValue, ~(uintptr_t)J9_GC_CONTINUATION_CONCURRENTSCANNING)) {
+			J9VMThread *carrierThread = (J9VMThread *) *localAddr;
+			omrthread_monitor_enter(carrierThread->publicFlagsMutex);
+			/*  notify the waiting carrierThread */
+			omrthread_monitor_notify(carrierThread->publicFlagsMutex);
+			omrthread_monitor_exit(carrierThread->publicFlagsMutex);
+		}
+	}
+#endif /* JAVA_SPEC_VERSION >= 19 */
+
 	static VMINLINE UDATA
-	walkContinuationStackFramesWrapper(J9VMThread *vmThread, j9object_t continuationObject, J9StackWalkState *walkState)
+	walkContinuationStackFramesWrapper(J9VMThread *vmThread, j9object_t continuationObject, J9StackWalkState *walkState, bool syncWithConcurrentGCScan)
 	{
 		UDATA rc = J9_STACKWALK_RC_NONE;
 #if JAVA_SPEC_VERSION >= 19
 		J9VMContinuation *continuation = J9VMJDKINTERNALVMCONTINUATION_VMREF(vmThread, continuationObject);
+		if (syncWithConcurrentGCScan && (NULL != continuation)) {
+			if (!tryEnterConcurrentContinuationMonitor(continuation)) {
+				return rc;
```

Comment:
GAC found it counterintuitive  that failing to 'enter monitor' ends up returning 'success' (RC_NONE). 
Besides, we don't have any explicit monitor, any more.
Perhaps a better name is tryWinning(Reserving?)ConcurrentGCScan? If so, the exit method should should be updated accordingly: exitConcurrentGCScan(Reservation?)

You could even repeat a lighter version of the fat comment associated with this method, just prior to return statement:
/* If continuation is mounted or already being scanned by another GC thread, we do nothing */

Reply:
Is this going to result in missing stack scans? A continuation always has a stack that needs walking, whether it's the stack for an unmounted continuation, or the stack of carrier thread for a mounted continuation.

Every continuation and every thread stack needs processing somewhere. I don't think it's valid to just do nothing in the case where the mount state is changing.

Reply:
Valid concern, but it should not result in missing scans. If continuation is not scanned by this GC thread now, it will be done during pre-mount or post-unmount (depending of concurrent style, SATB or incremental-update).
If pre-mount or post-mount does not occur during the concurrent phase then during initial STW GC increment root scan or during final STW increment when cleaning cards or re-scanning roots (again, depending on concurrent style)

This is why we use atomic: we must conclude that either we do here or it's done somewhere else, based on whether it's mounted or not. In this change we have to make sure this decision is made right (while we already made sure 'somewhere else' really does occur, as just explained).

Carrier thread are always scanned during root scanning, regardless if mounted something or not.